File: src/main/java/sample/design_pattern/Adapter.java
/*
 * Copyright 2013 SatoShun
 *
 * Adapter Pattern
 */

package sample.design_pattern;


interface IProduct {
    double getSize();
}


class Product implements IProduct {
    private double width;
    private double height;

    public Product(double width, double height) {
        this.width = width;
        this.height = height;
    }

    public double getSize() {
        return width * height;
    }
}


class NewProduct {
    private double width;
    private double height;
    private double mass;

    public NewProduct(double width, double height, double mass) {
        this.width = width;
        this.height = height;
        this.mass = mass;
    }

    public double getMass() {
        return width * height * mass;
    }
}


class AdapterProduct implements IProduct {
    private NewProduct product;

    public AdapterProduct(NewProduct product) {
        this.product = product;
    }

    public double getSize() {
        return product.getMass();
    }
}


class Adapter {
    public static void main(String[] args) {
        IProduct product = new Product(10, 20);
        showProduct(product);

        product = new AdapterProduct(new NewProduct(10, 20, 5));
        showProduct(product);
    }

    private static void showProduct(IProduct product) {
        System.out.println(product.getSize());
    }
}

File: src/main/scala/sample/design_pattern/Adapter.scala
/*
 * Copyright 2013 SatoShun
 *
 * Adapter Pattern
 */

package sample.design_pattern;


trait IProduct {
    def getSize(): Double;
}


class Product(val width: Double, val height: Double) extends IProduct {
    def getSize(): Double = {
        width * height
    }
}


class NewProduct(val width: Double, val height: Double, val mass: Double) {
    def getMass(): Double = {
        width * height * mass
    }
}

class AdapterProduct(val product: NewProduct) extends IProduct {
    def getSize(): Double = {
        product.getMass()
    }
}


object Adapter {
    def main(args: Array[String]) = {
        val adapter: IProduct = new AdapterProduct(
            new NewProduct(10.0D, 20.0D, 10.0D));
        showProduct(adapter)
        showProduct(new Product(10.0D, 30.0D))
    }

    private def showProduct(product: IProduct) = {
        println(product.getSize())
    }
}
